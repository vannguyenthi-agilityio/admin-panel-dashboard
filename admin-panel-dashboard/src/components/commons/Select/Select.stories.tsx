import type { Meta, StoryObj } from '@storybook/react';
import Select from '.';

const meta: Meta<typeof Select> = {
  args: {
    'aria-invalid': false,
    id: 'select-field',
    name: 'select-field'
  },
  argTypes: {
    'aria-invalid': {
      control: { type: 'boolean' },
    },
    'aria-selected': {
      control: { type: 'boolean' },
    },
    defaultValue: {
      type: 'string',
    },
    required: {
      type: 'boolean',
    },
    label: {
      type: 'string',
    },
    options: {
      control: { type: 'select' },
      options: [{
        option: {type: 'string'},
        value: {type: 'string'},
        id: {type: 'number'},
      }],
    },
  },
  component: Select,
  tags: ['autodocs'],
  title: 'Components/Commons/Select',
};

export default meta;
type Story = StoryObj<typeof meta>;

export const WithLabel: Story = {};

export const Label: Story = {
  args: {
    label: "Select Field",
  },
};

export const LabelRequired: Story = {
  args: {
    label: "Select Field",
    required: true,
  },
};
