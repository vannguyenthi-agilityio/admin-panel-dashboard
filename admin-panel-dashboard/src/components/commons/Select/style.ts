import { cva } from 'class-variance-authority';

export const selectStyles = cva(
  'w-full font-primary font-light text-sm bg-transparent placeholder:text-slate-400 text-slate-700 text-sm border border-slate-200 rounded-lg px-3 py-2 transition duration-300 ease focus:outline-none focus:border-slate-400 hover:border-slate-400 shadow-sm focus:shadow-md appearance-none cursor-pointer',
  {
    variants: {
      variant: {
        primary: ["bg-transparent", "text-dark"],
        secondary: ["bg-gray-100", "text-darker"],
      },
      selectSize: {
        xs: ["h-6", "text-xs", "px-2", "py-0"],
        sm: ["h-10", "text-sm", "px-2", "py-2"],
        md: ["h-12", "text-md", "px-2", "py-2"],
        lg: ["h-14", "text-lg", "px-2", "py-2"],
      },
    },
    defaultVariants: {
      variant: "primary",
      selectSize: "sm",
    },
  }
);
