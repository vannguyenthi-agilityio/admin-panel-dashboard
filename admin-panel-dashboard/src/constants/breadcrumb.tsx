// Types
import { IItemBreadcrumb } from "@/types";

// Constants
import { ROUTES } from "@/constants";

export const ITEM_BREADCRUMB: IItemBreadcrumb[] = [
  {
    href: ROUTES.HOME,
    crumbName: "DashBoard",
    linkActiveCrumb: false,
    linkIndex: 1
  },
  {
    href: ROUTES.CUSTOMERS,
    crumbName: "Customersâ€™ List",
    linkActiveCrumb: true,
    linkIndex: 2
  }
]

export const ITEM_BREADCRUMB_FORM: IItemBreadcrumb[] = [
  {
    href: ROUTES.HOME,
    crumbName: "DashBoard",
    linkActiveCrumb: false,
    linkIndex: 1
  },
  {
    href: ROUTES.CUSTOMERS,
    crumbName: "Customersâ€™ List",
    linkActiveCrumb: false,
    linkIndex: 2
  },
  {
    href: `${ROUTES.CUSTOMERS}/form`,
    crumbName: "New Customer",
    linkActiveCrumb: true,
    linkIndex: 3
  }
]

export const ICON_ITEM_BREADCRUMB: IItemBreadcrumb[] = [
  {
    href: ROUTES.HOME,
    crumbName: "DashBoard",
    linkActiveCrumb: false,
    linkIndex: 1,
    iconElement: <span className="mr-1">ðŸ‘‰</span>,
  },
  {
    href: ROUTES.CUSTOMERS,
    crumbName: "Customersâ€™ List",
    linkActiveCrumb: true,
    linkIndex: 2,
    iconElement: <span className="mr-1">ðŸ‘‰</span>,
  }
]
