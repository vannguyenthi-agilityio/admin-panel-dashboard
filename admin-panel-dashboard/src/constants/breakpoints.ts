export const BREAKPOINTS = {
  MD: 768,
  XL: 1280,
};
