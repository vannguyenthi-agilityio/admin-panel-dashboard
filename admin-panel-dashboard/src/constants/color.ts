export const COLOR = {
  PRIMARY: {
    50: '#fff'
  },
  SECONDARY: {
    50: '#000',
    100: '#adb5bd',
    200: '#212529'
  },
}
