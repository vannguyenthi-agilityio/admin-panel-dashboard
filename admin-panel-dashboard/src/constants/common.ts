// Types
import { OptionType } from "@/types";

export const MONTHS: OptionType[] = [
  {
    option: "January",
    value: "0",
  },
  {
    option: "February",
    value: "1",
  },
  {
    option: "March",
    value: "2",
  },
  {
    option: "April",
    value: "3",
  },
  {
    option: "May",
    value: "4",
  },
  {
    option: "June",
    value: "5",
  },
  {
    option: "July",
    value: "6",
  },
  {
    option: "August",
    value: "7",
  },
  {
    option: "September",
    value: "8",
  },
  {
    option: "October",
    value: "9",
  },
  {
    option: "November",
    value: "10",
  },
  {
    option: "December",
    value: "11",
  },
];
