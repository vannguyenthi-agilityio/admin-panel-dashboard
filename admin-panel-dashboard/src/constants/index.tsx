export * from './color';
export * from "./common";
