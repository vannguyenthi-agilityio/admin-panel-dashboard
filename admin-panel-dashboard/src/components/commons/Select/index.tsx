
import { memo, forwardRef } from "react";

import {
  ArrowDownIcon,
} from '@/components/Icons';

import { MONTHS } from "@/constants";

// Types
import { ISelect, OptionType } from "@/types";

// Helpers
import { clsxMerge } from '@/helpers';

import { selectStyles } from './style';

const Select =  forwardRef<HTMLSelectElement, ISelect>((
  { required,
    label,
    name,
    options = MONTHS,
    className,
    variant,
    selectSize,
    onChange,
    ...props }, ref) => {
  const classNameLabel = required
    ? "text-grey text-sm dark:text-lighter capitalize after:content-['*'] after:text-red-500 after:text-md"
    : "text-grey text-sm dark:text-lighter capitalize";
  return (
    <div className="w-full">
     {label && <label className={classNameLabel}>{label}</label>}
      <div className="relative">
        <select
          className={clsxMerge(
            selectStyles({ variant, selectSize }),
            className
          )}
          id={name}
          data-testid={name}
          ref={ref}
          onChange={onChange}
          {...props}
        >
          {options.map((item: OptionType) => (
            <option
              key={item.value}
              value={item.value.toString()}
              data-testid={`month-${item.value}`}
              className="cursor-pointer"
            >
              {item.option}
            </option>
            ))}
        </select>
        <ArrowDownIcon  className="ml-1 absolute top-[18px] right-3 text-slate-700"/>
    </div>
  </div>
  );
});

Select.displayName = "Select";
export default memo(Select);
