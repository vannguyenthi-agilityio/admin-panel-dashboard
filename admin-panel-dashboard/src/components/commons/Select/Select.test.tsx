import {
  render,
  fireEvent,
  waitFor,
  queryAllByAttribute
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Select from '.'

import { MONTHS } from "@/constants";

describe('Select component', () => {
  it('Renders select component', () => {
    const {container} = render(<Select
        label='Select field'
        options={MONTHS}
      />
    )
    expect(container).toMatchSnapshot();
  });

  it("Renders correct number of options", () => {
    const getById = queryAllByAttribute.bind(null, "id");
    const { container } = render(
      <Select options={MONTHS} name="month" />,
    );
    const selectElement = getById(container, "month");

    expect(selectElement).toBeTruthy();
    
    expect(selectElement[0]?.children).toHaveLength(12);
  });

   it("Renders correct name of options", () => {
    const { getAllByText } = render(
      <Select label="Month" options={MONTHS} defaultValue="4"/>,
    );
    expect(getAllByText("May")).toBeTruthy();
  });

  it("Should call onChange handler with selected value", async () => {
    const onChangeMock = jest.fn(); // Mock the onChange handler

    const { getByTestId } = render(
      <Select
        label="Test Select"
        options={MONTHS}
        onChange={onChangeMock}
        name="month"
      />,
    );

    fireEvent.click(getByTestId("month").firstElementChild as HTMLElement);
    fireEvent.click(getByTestId("month-2"));

    waitFor(() => {
      expect(onChangeMock).toHaveBeenCalledTimes(1);
      expect(onChangeMock).toHaveBeenCalledWith(MONTHS[2].value);
    });
  });
})
