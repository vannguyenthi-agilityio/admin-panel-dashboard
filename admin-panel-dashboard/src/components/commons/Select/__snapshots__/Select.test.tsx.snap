// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component Renders select component 1`] = `
<div>
  <div
    class="w-full"
  >
    <label
      class="text-grey text-sm dark:text-lighter capitalize"
    >
      Select field
    </label>
    <div
      class="relative"
    >
      <select
        class="w-full font-primary font-light placeholder:text-slate-400 border border-slate-200 rounded-lg transition duration-300 ease focus:outline-none focus:border-slate-400 hover:border-slate-400 shadow-sm focus:shadow-md appearance-none cursor-pointer bg-transparent text-dark h-10 text-sm px-2 py-2"
      >
        <option
          class="cursor-pointer"
          data-testid="month-0"
          value="0"
        >
          January
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-1"
          value="1"
        >
          February
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-2"
          value="2"
        >
          March
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-3"
          value="3"
        >
          April
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-4"
          value="4"
        >
          May
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-5"
          value="5"
        >
          June
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-6"
          value="6"
        >
          July
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-7"
          value="7"
        >
          August
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-8"
          value="8"
        >
          September
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-9"
          value="9"
        >
          October
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-10"
          value="10"
        >
          November
        </option>
        <option
          class="cursor-pointer"
          data-testid="month-11"
          value="11"
        >
          December
        </option>
      </select>
      <svg
        class="ml-1 absolute top-[18px] right-3 text-slate-700"
        fill="none"
        height="5"
        viewBox="0 0 8 5"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.25 0.75L4 4.25L0.75 0.75"
          stroke="#adb5bd"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </div>
  </div>
</div>
`;
